File: backend/src/modules/moderation/ruleEngine.test.ts
import { describe, expect, it } from "vitest";
import { evaluateMessage, type ChatMessage, type ModerationProfile } from "./ruleEngine.js";

const profile: ModerationProfile = {
  blockedTerms: ["scam", "cheap views"],
  regexPatterns: [],
  linkPolicy: "delete",
  allowedDomains: [],
  blockedDomains: [],
  capsThreshold: 0.75,
  maxRepeatedCharacters: 5,
  maxEmojiCount: 8,
  maxMentions: 6,
  maxSymbolCount: 16,
  trustedChannelIds: [],
  temporaryTrustedChannels: [],
  ignoreMembers: false,
  raidMode: false,
  newChatterBurstThreshold: 6,
  newChatterBurstWindowSeconds: 30,
  firstStreamMinutesOnly: null,
  autoReplyEnabled: false,
  autoReplyMessage: "Please keep chat safe and on topic.",
  hideUserOnSevereMatch: false
};

function message(text: string, overrides: Partial<ChatMessage> = {}): ChatMessage {
  return {
    id: "message-1",
    authorChannelId: "viewer-1",
    authorName: "Viewer",
    text,
    timestamp: new Date().toISOString(),
    ...overrides
  };
}

describe("evaluateMessage", () => {
  it("deletes messages with blocked terms", () => {
    const result = evaluateMessage(message("This is a scam"), profile);

    expect(result.matched).toBe(true);
    expect(result.actions).toContainEqual(
      expect.objectContaining({
        type: "deleteMessage",
        reason: "blocked_term:scam"
      })
    );
  });

  it("applies the configured link policy", () => {
    const result = evaluateMessage(message("check https://example.com"), profile);

    expect(result.actions).toContainEqual(
      expect.objectContaining({
        type: "deleteMessage",
        reason: "link_policy"
      })
    );
  });

  it("flags excessive caps without deleting by default", () => {
    const result = evaluateMessage(message("THIS IS TOO LOUD"));

    expect(result.actions).toContainEqual(
      expect.objectContaining({
        type: "flagForReview",
        reason: "excessive_caps"
      })
    );
  });

  it("allows channel owners and moderators", () => {
    const result = evaluateMessage(message("scam link", { isModerator: true }), profile);

    expect(result.matched).toBe(false);
    expect(result.actions).toEqual([
      {
        type: "allow",
        reason: "trusted_author",
        confidence: 1
      }
    ]);
  });

  it("deletes messages matching custom regex patterns", () => {
    const result = evaluateMessage(
      message("free nitro giveaway"),
      {
        ...profile,
        regexPatterns: ["free\\s+nitro"]
      }
    );

    expect(result.actions).toContainEqual(
      expect.objectContaining({
        type: "deleteMessage",
        reason: "regex_pattern:1"
      })
    );
  });

  it("ignores invalid regex patterns", () => {
    const result = evaluateMessage(
      message("free nitro giveaway"),
      {
        ...profile,
        regexPatterns: ["["]
      }
    );

    expect(result.matched).toBe(false);
  });

  it("deletes explicitly blocked domains", () => {
    const result = evaluateMessage(
      message("check https://bad.example/path"),
      {
        ...profile,
        linkPolicy: "allow",
        blockedDomains: ["example"]
      }
    );

    expect(result.actions).toContainEqual(
      expect.objectContaining({
        type: "deleteMessage",
        reason: "blocked_domain:bad.example"
      })
    );
  });

  it("flags links outside the allowlist", () => {
    const result = evaluateMessage(
      message("watch www.notallowed.test"),
      {
        ...profile,
        linkPolicy: "flag",
        allowedDomains: ["youtube.com"]
      }
    );

    expect(result.actions).toContainEqual(
      expect.objectContaining({
        type: "flagForReview",
        reason: "domain_not_allowed:notallowed.test"
      })
    );
  });

  it("flags emoji and mention spam", () => {
    const result = evaluateMessage(
      message("@a @b @c 😀😀😀"),
      {
        ...profile,
        maxEmojiCount: 2,
        maxMentions: 2
      }
    );

    expect(result.actions).toContainEqual(expect.objectContaining({ reason: "emoji_spam" }));
    expect(result.actions).toContainEqual(expect.objectContaining({ reason: "mention_spam" }));
  });

  it("flags symbol spam", () => {
    const result = evaluateMessage(
      message("%%%% $$$$ #### !!!! @@@@"),
      {
        ...profile,
        maxSymbolCount: 8
      }
    );

    expect(result.actions).toContainEqual(expect.objectContaining({ reason: "symbol_spam" }));
  });

  it("applies raid mode stricter thresholds", () => {
    const result = evaluateMessage(
      message("THIS IS LOUD"),
      {
        ...profile,
        capsThreshold: 0.9,
        raidMode: true
      }
    );

    expect(result.actions).toContainEqual(expect.objectContaining({ reason: "excessive_caps" }));
  });

  it("can trust channel members when configured", () => {
    const result = evaluateMessage(
      message("this is a scam", { isMember: true }),
      {
        ...profile,
        ignoreMembers: true
      }
    );

    expect(result.matched).toBe(false);
    expect(result.actions).toEqual([
      {
        type: "allow",
        reason: "trusted_member",
        confidence: 1
      }
    ]);
  });

  it("allows verified YouTube authors before applying filters", () => {
    const result = evaluateMessage(
      message("this is a scam", { isVerified: true }),
      profile
    );

    expect(result.matched).toBe(false);
    expect(result.actions).toEqual([
      {
        type: "allow",
        reason: "trusted_verified",
        confidence: 1
      }
    ]);
  });

  it("allows whitelisted channel IDs before applying filters", () => {
    const result = evaluateMessage(
      message("this is a scam", { authorChannelId: "trusted-viewer" }),
      {
        ...profile,
        trustedChannelIds: ["trusted-viewer"]
      }
    );

    expect(result.matched).toBe(false);
    expect(result.actions).toEqual([
      {
        type: "allow",
        reason: "trusted_channel",
        confidence: 1
      }
    ]);
  });

  it("allows unexpired temporary trusted channel IDs before applying filters", () => {
    const result = evaluateMessage(
      message("this is a scam", { authorChannelId: "temp-viewer" }),
      {
        ...profile,
        temporaryTrustedChannels: [
          {
            channelId: "temp-viewer",
            expiresAt: new Date(Date.now() + 60_000).toISOString()
          }
        ]
      }
    );

    expect(result.matched).toBe(false);
    expect(result.actions).toEqual([
      {
        type: "allow",
        reason: "temporary_trusted_channel",
        confidence: 1
      }
    ]);
  });

  it("ignores expired temporary trusted channel IDs", () => {
    const result = evaluateMessage(
      message("this is a scam", { authorChannelId: "temp-viewer" }),
      {
        ...profile,
        temporaryTrustedChannels: [
          {
            channelId: "temp-viewer",
            expiresAt: new Date(Date.now() - 60_000).toISOString()
          }
        ]
      }
    );

    expect(result.matched).toBe(true);
    expect(result.actions).toContainEqual(
      expect.objectContaining({
        reason: "blocked_term:scam"
      })
    );
  });

  it("adds a configured auto-reply action when moderation rules match", () => {
    const result = evaluateMessage(
      message("this is a scam"),
      {
        ...profile,
        autoReplyEnabled: true,
        autoReplyMessage: "Please keep chat clean."
      }
    );

    expect(result.actions).toContainEqual(
      expect.objectContaining({
        type: "sendAutoReply",
        reason: "auto_reply",
        text: "Please keep chat clean."
      })
    );
  });

  it("does not add an auto-reply for trusted users", () => {
    const result = evaluateMessage(
      message("this is a scam", { isModerator: true }),
      {
        ...profile,
        autoReplyEnabled: true
      }
    );

    expect(result.matched).toBe(false);
    expect(result.actions).not.toContainEqual(
      expect.objectContaining({
        type: "sendAutoReply"
      })
    );
  });

  it("adds an opt-in hide-user action for severe rule matches", () => {
    const result = evaluateMessage(
      message("this is a scam"),
      {
        ...profile,
        hideUserOnSevereMatch: true
      }
    );

    expect(result.actions).toContainEqual(
      expect.objectContaining({
        type: "hideUser",
        reason: "severe_match_hide_user"
      })
    );
  });

  it("does not hide trusted users even when severe-match hiding is enabled", () => {
    const result = evaluateMessage(
      message("this is a scam", { isVerified: true }),
      {
        ...profile,
        hideUserOnSevereMatch: true
      }
    );

    expect(result.matched).toBe(false);
    expect(result.actions).toEqual([
      {
        type: "allow",
        reason: "trusted_verified",
        confidence: 1
      }
    ]);
  });

  it("applies rules during the configured first stream minutes", () => {
    const result = evaluateMessage(
      message("this is a scam", {
        timestamp: "2026-06-07T18:05:00.000Z",
        streamStartedAt: "2026-06-07T18:00:00.000Z"
      }),
      {
        ...profile,
        firstStreamMinutesOnly: 10
      }
    );

    expect(result.matched).toBe(true);
    expect(result.actions).toContainEqual(
      expect.objectContaining({
        reason: "blocked_term:scam"
      })
    );
  });

  it("skips rules outside the configured first stream minutes", () => {
    const result = evaluateMessage(
      message("this is a scam", {
        timestamp: "2026-06-07T18:11:00.000Z",
        streamStartedAt: "2026-06-07T18:00:00.000Z"
      }),
      {
        ...profile,
        firstStreamMinutesOnly: 10
      }
    );

    expect(result.matched).toBe(false);
    expect(result.actions).toEqual([
      {
        type: "allow",
        reason: "outside_first_stream_window",
        confidence: 1
      }
    ]);
  });
});

File: android/app/src/test/java/com/chatmod/mobile/domain/rules/RuleEngineTest.kt
package com.chatmod.mobile.domain.rules

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant

class RuleEngineTest {
    private val engine = RuleEngine()
    private val profile = ModerationProfile(
        blockedTerms = listOf("scam", "cheap views"),
        linkPolicy = LinkPolicy.Delete,
        capsThreshold = 0.75,
        maxRepeatedCharacters = 5
    )

    @Test
    fun blockedTermsDeleteMessages() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "this is a scam"),
            profile
        )

        assertTrue(decision.matched)
        assertTrue(decision.actions.any { it.type == ActionType.DeleteMessage && it.reason == "blocked_term:scam" })
    }

    @Test
    fun moderatorsAreTrusted() {
        val decision = engine.evaluate(
            ChatMessage("m1", "mod-1", "Mod", "scam link", isModerator = true),
            profile
        )

        assertEquals(false, decision.matched)
        assertEquals(ActionType.Allow, decision.actions.single().type)
    }

    @Test
    fun customRegexPatternsDeleteMessages() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "free nitro giveaway"),
            profile.copy(regexPatterns = listOf("""free\s+nitro"""))
        )

        assertTrue(decision.actions.any { it.type == ActionType.DeleteMessage && it.reason == "regex_pattern:1" })
    }

    @Test
    fun invalidRegexPatternsAreIgnored() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "free nitro giveaway"),
            profile.copy(regexPatterns = listOf("["))
        )

        assertEquals(false, decision.matched)
    }

    @Test
    fun domainBlocklistDeletesLinkedMessages() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "check https://bad.example/path"),
            profile.copy(linkPolicy = LinkPolicy.Allow, blockedDomains = listOf("example"))
        )

        assertTrue(decision.actions.any { it.type == ActionType.DeleteMessage && it.reason == "blocked_domain:bad.example" })
    }

    @Test
    fun domainAllowlistFlagsUnknownDomains() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "watch www.notallowed.test"),
            profile.copy(linkPolicy = LinkPolicy.Flag, allowedDomains = listOf("youtube.com"))
        )

        assertTrue(decision.actions.any { it.type == ActionType.FlagForReview && it.reason == "domain_not_allowed:notallowed.test" })
    }

    @Test
    fun emojiAndMentionSpamAreFlagged() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "@a @b @c 😀😀😀"),
            profile.copy(maxEmojiCount = 2, maxMentions = 2)
        )

        assertTrue(decision.actions.any { it.reason == "emoji_spam" })
        assertTrue(decision.actions.any { it.reason == "mention_spam" })
    }

    @Test
    fun symbolSpamIsFlagged() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "%%%% $$$$ #### !!!! @@@@"),
            profile.copy(maxSymbolCount = 8)
        )

        assertTrue(decision.actions.any { it.type == ActionType.FlagForReview && it.reason == "symbol_spam" })
    }

    @Test
    fun raidModeAppliesStricterThresholds() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "THIS IS LOUD"),
            profile.copy(capsThreshold = 0.90, raidMode = true)
        )

        assertTrue(decision.actions.any { it.type == ActionType.FlagForReview && it.reason == "excessive_caps" })
    }

    @Test
    fun configuredMembersAreTrusted() {
        val decision = engine.evaluate(
            ChatMessage("m1", "member-1", "Member", "this is a scam", isMember = true),
            profile.copy(ignoreMembers = true)
        )

        assertEquals(false, decision.matched)
        assertEquals("trusted_member", decision.actions.single().reason)
    }

    @Test
    fun trustedChannelIdsAreAllowedBeforeFilters() {
        val decision = engine.evaluate(
            ChatMessage("m1", "trusted-viewer", "Trusted Viewer", "this is a scam"),
            profile.copy(trustedChannelIds = listOf("trusted-viewer"))
        )

        assertEquals(false, decision.matched)
        assertEquals("trusted_channel", decision.actions.single().reason)
    }

    @Test
    fun verifiedAuthorsAreAllowedBeforeFilters() {
        val decision = engine.evaluate(
            ChatMessage("m1", "verified-viewer", "Verified Viewer", "this is a scam", isVerified = true),
            profile
        )

        assertEquals(false, decision.matched)
        assertEquals("trusted_verified", decision.actions.single().reason)
    }

    @Test
    fun temporaryTrustedChannelIdsAreAllowedBeforeFilters() {
        val decision = engine.evaluate(
            ChatMessage("m1", "temp-viewer", "Temp Viewer", "this is a scam"),
            profile.copy(
                temporaryTrustedChannels = listOf(
                    TemporaryTrustedChannel(
                        channelId = "temp-viewer",
                        expiresAt = Instant.now().plusSeconds(60).toString()
                    )
                )
            )
        )

        assertEquals(false, decision.matched)
        assertEquals("temporary_trusted_channel", decision.actions.single().reason)
    }

    @Test
    fun configuredAutoReplyAddsReplyActionForMatchedRules() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "this is a scam"),
            profile.copy(
                autoReplyEnabled = true,
                autoReplyMessage = "Please keep chat clean."
            )
        )

        assertTrue(
            decision.actions.any {
                it.type == ActionType.SendAutoReply &&
                    it.reason == "auto_reply" &&
                    it.text == "Please keep chat clean."
            }
        )
    }

    @Test
    fun configuredAutoReplyDoesNotReplyToTrustedUsers() {
        val decision = engine.evaluate(
            ChatMessage("m1", "mod-1", "Mod", "this is a scam", isModerator = true),
            profile.copy(autoReplyEnabled = true)
        )

        assertEquals(false, decision.matched)
        assertTrue(decision.actions.none { it.type == ActionType.SendAutoReply })
    }

    @Test
    fun severeMatchHidingAddsHideUserActionWhenEnabled() {
        val decision = engine.evaluate(
            ChatMessage("m1", "viewer-1", "Viewer", "this is a scam"),
            profile.copy(hideUserOnSevereMatch = true)
        )

        assertTrue(
            decision.actions.any {
                it.type == ActionType.HideUser &&
                    it.reason == "severe_match_hide_user"
            }
        )
    }

    @Test
    fun severeMatchHidingDoesNotHideTrustedUsers() {
        val decision = engine.evaluate(
            ChatMessage("m1", "verified-viewer", "Verified Viewer", "this is a scam", isVerified = true),
            profile.copy(hideUserOnSevereMatch = true)
        )

        assertEquals(false, decision.matched)
        assertEquals("trusted_verified", decision.actions.single().reason)
        assertTrue(decision.actions.none { it.type == ActionType.HideUser })
    }

    @Test
    fun firstStreamMinutesTargetingAppliesRulesInsideWindow() {
        val decision = engine.evaluate(
            ChatMessage(
                id = "m1",
                authorChannelId = "viewer-1",
                authorName = "Viewer",
                text = "this is a scam",
                timestampMillis = 5 * 60_000L,
                streamStartedAtMillis = 0L
            ),
            profile.copy(firstStreamMinutesOnly = 10)
        )

        assertTrue(decision.matched)
        assertTrue(decision.actions.any { it.reason == "blocked_term:scam" })
    }

    @Test
    fun firstStreamMinutesTargetingSkipsRulesOutsideWindow() {
        val decision = engine.evaluate(
            ChatMessage(
                id = "m1",
                authorChannelId = "viewer-1",
                authorName = "Viewer",
                text = "this is a scam",
                timestampMillis = 11 * 60_000L,
                streamStartedAtMillis = 0L
            ),
            profile.copy(firstStreamMinutesOnly = 10)
        )

        assertEquals(false, decision.matched)
        assertEquals("outside_first_stream_window", decision.actions.single().reason)
    }

    @Test
    fun expiredTemporaryTrustedChannelIdsAreIgnored() {
        val decision = engine.evaluate(
            ChatMessage("m1", "temp-viewer", "Temp Viewer", "this is a scam"),
            profile.copy(
                temporaryTrustedChannels = listOf(
                    TemporaryTrustedChannel(
                        channelId = "temp-viewer",
                        expiresAt = Instant.now().minusSeconds(60).toString()
                    )
                )
            )
        )

        assertTrue(decision.matched)
        assertTrue(decision.actions.any { it.reason == "blocked_term:scam" })
    }
}
